src/libopensc/card-piv.c

```diff
@@ -3355,21 +3428,60 @@ static int piv_logout(sc_card_t *card)
 }
 
 
+/*
+ * Called when a sc_lock gets a reader lock and PCSC SCardBeginTransaction
+ * If SCardBeginTransaction may pass back tha a card reset was seen since
+ * the last transaction  completed.
+ * There may have been one or more resets, by other card drivers in different
+ * processes, and they may have taken action already
+ * and changed the AID and or may have sent a  VERIFY with PIN
+ * So test the state of the card.
+ * this is very similiar to what the piv_match routtine does,
+ */
+
 static int piv_card_reader_lock_obtained(sc_card_t *card, int was_reset)
 {
 	int r = 0;
-	u8 temp[2000];
+	u8 temp[256];
 	size_t templen = sizeof(temp);
+	struct sc_pin_cmd_data data;
 	piv_private_data_t * priv = PIV_DATA(card); /* may be null */
 
 	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
-	if (was_reset > 0) {
-		if (priv)
-			priv->logged_in =  SC_PIN_STATE_UNKNOWN;
 
-		r = piv_select_aid(card, piv_aids[0].value, piv_aids[0].len_short, temp, &templen);
+	/* We have a PCSC transaction and sc_lock */
+	if (priv == NULL || priv->pstate == PIV_STATE_MATCH) {
+		sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,
+			    priv ? "PIV_STATE_MATCH" : "priv==NULL");
+		r = 0; /* do nothing, piv_match will take care of it */
+		goto err;
 	}
 
+	/* make sure our application is active */
+
+	/* first see if AID is active AID be reading discovery obkect '7E' */
+	/* If not try selecting AID */
+	r = piv_find_discovery(card);
+	if (r < 0) 
+		r = piv_select_aid(card, piv_aids[0].value, piv_aids[0].len_short, temp, &templen);
+
+	if (r < 0) /* bad error return will show up in sc_lock as error*/
+		goto err;
+	
+	if (was_reset > 0)
+		priv->logged_in =  SC_PIN_STATE_UNKNOWN;
+
+	/* See if VERIFY Lc=empty will tell us the state */
+	memset(&data, 0, sizeof(data));
+	data.cmd = SC_PIN_CMD_GET_INFO;
+	data.pin_type = SC_AC_CHV;
+	data.pin_reference =  priv->pin_preference;
+	/* will try our best to see if logged_in or not */
+	r = piv_pin_cmd(card, &data, NULL);
```

Comment:
unused assignment.

Why do you need to check for the PIN at all? Maybe only in the case of a reset?

Reply:
That is trying to keep the card->logged_in state correct. With the Yubico it is easy to lose the state. 
For example: Assuming disconnect=leave
Current process is using PIV application and logged it. 
Other process started and some driver sets the AID to something else, and then sets it to 
PIV.  
Our process uses discovery  (7E) to see the PIV is active, but it does not know if the login state has changed unless it tests. The is the test.

Reply:
I think this command is not needed. Simply set the PIV's internal tracking to SC_PIN_STATE_UNKNOWN if the applet has been deselected or reset. An application that wants to know the PIN status directly or indirectly calls `sc_pin_cmd` with `SC_PIN_CMD_GET_INFO`; this issues the VERIFY command if the card supports it.

Note that we should try to remove the driver specific tracking and put this into the PKCS#11 specific tracking. (i.e. we need to propagate a reset, for example). We should not do a double entry-bookeeping.

Reply:
OK, will look at that closer.

Reply:
Did you try [`piv_card_reader_lock_obtained`](https://github.com/OpenSC/OpenSC/pull/1256/files#r164997828) without the PIN status query? I still think that's an superfluous APDU, but if you think that's necessary then please remove the unused assignment (which static code analysis will complain about).